cmd/kube-proxy/app/server.go

```diff
@@ -1063,3 +1064,28 @@ func getNodeIPs(logger klog.Logger, client clientset.Interface, name string) []n
 	}
 	return nodeIPs
 }
+
+func (o *Options) setDefaultAddresses(config *kubeproxyconfig.KubeProxyConfiguration) {
+	defaultHealthzAddress, defaultMetricsAddress := getDefaultAddresses(config.BindAddress)
```

Comment:
The config's usage has declared the current behavior:
> --bind-address string     Default: 0.0.0.0
>
>   Overrides kube-proxy's idea of what its node's primary IP is. Note that the name is a historical artifact, and kube-proxy does not actually bind any sockets to this IP.

This change may even break users who correctly use `--bind-address` to specify node's primary IP and `--healthz-bind-address` to specify HealthzBindAddress.

Reply:
ok, there is also another usage declaring the behavior you are fixing:
> --healthz-bind-address ipport     Default: 0.0.0.0:10256
>
> The IP address and port for the health check server to serve on, defaulting to "0.0.0.0:10256" (if --bind-address is unset or IPv4), or "[::]:10256" (if --bind-address is IPv6). Set empty to disable. This parameter is ignored if a config file is specified by --config.

Reply:
Yes, this is the key. This PR solution fits this description

Reply:
I thought this change would break "Set empty to disable" if it overrides the addresses after defaulting. However, I then found setting empty value didn't work even before the patch as an explicit `--healthz-bind-address=` can't reset the default value either. Not sure if this is a doc issue, or empty value is really expected to disable it. If the latter, perhaps moving defaulting here may make it harder to be disabled?

As a side note, setting "bindAddress" to ipv6 address in config file does make metrics and healthz listen to IPv6 addresses. https://github.com/kubernetes/kubernetes/pull/120274 updated the usages in both config file and CLI flags, it's also possible that it should have updated config file only? cc @danwinship

Reply:
> As a side note, setting "bindAddress" to ipv6 address in config file does make metrics and healthz listen to IPv6 addresses. #120274 updated the usages in both config file and CLI flags, it's also possible that it should have updated config file only? cc @danwinship

Ah... yeah, I guess that's probably the real bug here; I assumed that the CLI options had the same semantics as the config options here, but actually they didn't.

At this point, I think we should not be changing the semantics of the CLI options, even where they are clearly bad. The plan is to deprecate them in favor of the v1alpha2/eventual-v1 config file anyway.